native/dggsvp3.go

```diff
@@ -0,0 +1,273 @@
+		}
+		impl.Dorg2r(p, p, min(p, n), v, ldv, tau, work)
+	}
+
+	// Clean up B.
+	for i := 1; i < l; i++ {
+		r := b[i*ldb : i*ldb+i]
+		for j := range r {
+			r[j] = 0
+		}
+	}
+	if p > l {
+		impl.Dlaset(blas.All, p-l, n, 0, 0, b[l*ldb:], ldb)
+	}
+
+	if wantq {
+		// Set Q = I and update Q := Q*P.
+		impl.Dlaset(blas.All, n, n, 0, 1, q, ldq)
+		impl.Dlapmt(forward, n, n, q, ldq, iwork)
+	}
+
+	if p >= l && n != l {
+		// RQ factorization of [ S11 S12 ]: [ S11 S12 ] = [ 0 S12 ]*Z.
+		impl.Dgerq2(l, n, b, ldb, tau, work)
+
+		// Update A := A*Z^T.
+		impl.Dormr2(blas.Right, blas.Trans, m, n, l, b, ldb, tau, a, lda, work)
+
+		if wantq {
+			// Update Q := Q*Z^T.
+			impl.Dormr2(blas.Right, blas.Trans, n, n, l, b, ldb, tau, q, ldq, work)
+		}
+
+		// Clean up B.
+		impl.Dlaset(blas.All, l, n-l, 0, 0, b, ldb)
+		for i := 1; i < l; i++ {
+			r := b[i*ldb+n-l : i*ldb+i+n-l]
+			for j := range r {
+				r[j] = 0
+			}
+		}
+	}
+
+	// Let              N-L     L
+	//            A = [ A11    A12 ] M,
+	//
+	// then the following does the complete QR decomposition of A11:
+	//
+	//          A11 = U*[  0  T12 ]*P1^T.
+	//                  [  0   0  ]
+	for i := range iwork[:n-l] {
+		iwork[i] = 0
+	}
+	impl.Dgeqp3(m, n-l, a, lda, iwork[:n-l], tau, work, lwork)
+
+	// Determine the effective rank of A11.
+	for i := 0; i < min(m, n-l); i++ {
+		if math.Abs(a[i*lda+i]) > tola {
+			k++
+		}
+	}
+
+	// Update A12 := U^T*A12, where A12 = A[0:m, n-l:n].
+	impl.Dorm2r(blas.Left, blas.Trans, m, l, min(m, n-l), a, lda, tau, a[n-l:], lda, work)
+
+	if wantu {
+		// Copy the details of U, and form U.
+		impl.Dlaset(blas.All, m, m, 0, 0, u, ldu)
+		if m > 1 {
+			impl.Dlacpy(blas.Lower, m-1, min(m, n-l), a[lda:], lda, u[ldu:], ldu)
+		}
+		impl.Dorg2r(m, m, min(m, n-l), u, ldu, tau, work)
+	}
+
+	if wantq {
+		// Update Q[0:n, 0:n-l] := Q[0:n, 0:n-l]*P1.
+		impl.Dlapmt(forward, n, n-l, q, ldq, iwork[:n-l])
+	}
+
+	// Clean up A: set the strictly lower triangular part of
+	// A[0:k, 0:k] = 0, and A[k:m, 0:n-l] = 0.
+	for i := 1; i < k; i++ {
+		r := a[i*lda : i*lda+i]
+		for j := range r {
+			r[j] = 0
+		}
+	}
+	if m > k {
+		impl.Dlaset(blas.All, m-k, n-l, 0, 0, a[k*lda:], lda)
+	}
+
+	if n-l > k {
+		// RQ factorization of [ T11 T12 ] = [ 0 T12 ]*Z1.
+		impl.Dgerq2(k, n-l, a, lda, tau, work)
+
+		if wantq {
+			// Update Q[0:n, 0:n-l] := Q[0:n, 0:n-l]*Z1^T.
+			impl.Dorm2r(blas.Right, blas.Trans, n, n-l, k, a, lda, tau, q, ldq, work)
+		}
+
+		// Clean up A.
+		impl.Dlaset(blas.All, k, n-l-k, 0, 0, a, lda)
+		for i := 1; i < k; i++ {
+			r := a[i*lda+n-k-l : i*lda+i+n-k-l]
+			for j := range r {
+				a[j] = 0
+			}
+		}
+	}
+
+	if m > k {
+		// QR factorization of A[k:m, n-l:n].
+		impl.Dgeqr2(m-k, l, a[k*lda+n-l:], lda, tau, work)
+		if wantu {
+			// Update U[:, k:m) := U[:, k:m]*U1.
+			impl.Dorm2r(blas.Right, blas.NoTrans, m, m-k, min(m-k, l), a[k*lda+n-l:], lda, tau, u[k:], ldu, work)
+		}
+
+		// Clean up A.
+		for i := k + 1; i < n; i++ {
```

Comment:
The solution to this and the issue below is to range up to `k+l`. Then we are limiting our rows to A23 and the offset is correct for being with the columns of that block.